Bitbucket.Authentication/Authority.cs

```diff
@@ -77,7 +74,35 @@ public string UserUrl
                 OAuth.OAuthAuthenticator oauth = new OAuth.OAuthAuthenticator();
                 try
                 {
-                    return await oauth.GetAuthAsync(targetUri, scope, CancellationToken.None);
+                    var result = await oauth.GetAuthAsync(targetUri, scope, CancellationToken.None);
+
+                    if (!result.IsSuccess)
+                    {
+                        Trace.WriteLine($"oauth authentication failed");
+                        return new AuthenticationResult(AuthenticationResultType.Failure);
+                    }
+
+                    // we got a toke but lets check to see the usernames match
+                    var restRootUri = new Uri(_restRootUrl);
+                    var authHeader = GetBearerHeaderAuthHeader(result.Token.Value);
+                    var userResult = await RestClient.TryGetUser(targetUri, RequestTimeout, restRootUri, authHeader);
+
+                    if(!userResult.IsSuccess)
+                    {
+                        Trace.WriteLine($"oauth user check failed");
+                        return new AuthenticationResult(AuthenticationResultType.Failure);
+                    }
+
+                    if (!string.IsNullOrWhiteSpace(userResult.RemoteUsername) && !username.Equals(userResult.RemoteUsername))
+                    {
```

Comment:
If the username stored in Bitbucket isn't the one supplied by the user, e.g. they used an email address, then keep hold of the Bitbucket one because we will store that for the credentials